README.md
# Roundcube-Vacation-Autoreply-plugin
Rebuild autoreply plugin for Roundcube, originaly from Peter Ruiter

This old plugin (but still useful) form 2010 need some modifications to be fully functional with Roundcube 1.2.3.

It's done !


Changelog
CHANGELOG for the Roundcube vacation plugin

Version 20161209
- [general] Transfered on github with all needed modifications to work with Roundcube 1.2.3, code changes by Yves Bernard <yves@lndm.fr>
- [virtual] Deleted MDB2 access (no more used since PHP5.3)
- [virutal] Added utf8_encode for subject and body at "get" (accentuated characters problem)
- [driver] Added utf8_decode for the same reason
- [general] Modified form generation (display problem)
- [template] Modified page content margins
- [template] Modified default template
- [translations] Added French translation
- [translations] Added Catalan translation
- [translations] Added Swedish translation
- [translations] Added Finnish translation
- [translations] Added Portuguese translation


****************

Version 20100613
- [documentation] Update INSTALL.TXT. config.inc.php refs removed
- [documentation] Default body/subject edited. 

Version 20100606
- [virtual] Added WHERE clause to update of Vacation table
- [dotforward] Rewrite of creating .forward files

Version 20100415
- [translations] Hungarian translation by zeta67
- [configuration] PHP < 5.3 needs double quotes in config.ini for parsing
- [dotforward] Use "|/usr/bin/vacation" rather than |"/usr/bin/vacation"

Version 20100305
- [all drivers] Ability to turn off setting forwards. Set disable_forward = True. 
- [Sieve] Mockup for now.
- [documentation] Passive mode is off by default.
- [configuration] Config.inc.php merged with config.ini. Config.inc.php is no longer needed
- [translations] Slovakian translation by Jan Vajda

Version 20100304
- [translations] Italian translation by Niccolo Rigacci
- [dotforward] Only keep a local copy if it's enabled by the end-user. Reported by Tim
- [vacationconfig] Specifies required settings per driver for config.ini

Version 20100227
- [Documentation] Rewrote documention in INSTALL.TXT and README.TXT
- [setuid] Aliases weren't used.
- [setuid] Check for correct user when executing setuid binary
- [virtual] Added better error checking after executing queries.
- [virtual] Updated vacation.pl to latest version
- [virtual] Support for always_keep_message and createvacationconf
- [sshftp/ftp/setuid] Disable 'Get aliases' button when there is only 1 identity

Version 20100222
- [template] Fixed styling, increased size for subject and body
- [translations] Seperate -> Separate.
- [ftp] References to .forward22 (for testing) removed.

Version 20100220
- [config] Overhauled configuration by using config.ini for improved multiple hosts/drivers

- [documentation] Updated documentation
- [template] Support for default subject and body (loaded from file). See config.inc.php
- [sshftp] Initial support for FTP over SSH with patches from Johnson Chow
- [template] Aliases can be retrieved using a button.
- [virtual] Check for domain lookup query returning 1 row.
- [dotforward] Alias localized and added to the localization/*.inc
- [dotforward] Keep a copy of the e-mail defaults to true now.
- [template] Allow user to enter aliases. Can be turned off in config.inc.php
- [general] Change error code to 601 so error is shown.
- [template] 'vacation' tab button in the prefs was not localized properly
- [translations] Taiwanese translation by Johnson Chow
- [translations] Polish translation by Rafal Domeracki
- [translations] Czech translation by xmacan
- [translations] Added Spanish translation

Version 20091008
- [setuid] always_keep_message config option keeps .vacation.msg when vacation is disabled
- [ftp] always_keep_message config option keeps .vacation.msg when vacation is disabled
- [setuid] deleting .forward files had a bug and did not quote the password
- Updated README

Version 20090927
- [setuid] class is working now
- [VacationDriverFactory] use file_exists and not is_class
- [ftp] improved _get() and downloadfile in case there is no .forward file
- [template] Fixed bug in height
- [sql] Add Postgresql schema

Version 20090918
- new configuration option for virtual: always_keep_copy. Always creates a %e -> %e alias
- simplied _get() array usage
- using placeholders for insert statement for vacation
- typo domainLookup that caused $this->domain_id to be empty

Version 20090917
- virtual_alias method simplied and rewritten to support comma seperated aliases
- virtual_alias method still supports old style one row per alias
- setVacation now saves a single row for for all aliases per person
  (compatibility for postfixadmin)
- NOTE: Incompatible config.inc.php change in delete_query
- Removed is_active() in virtual.class
- setVacation now keep's a copy when using auto-reply is set
- German translation provided by Sebastian Liechti

Version 20090916
- Virtual_alias now uses config.inc.php's select_query
- config.inc.php's select_query now needs the destination column rather than just 1

Version 20090913
- Fixed some typo's in various files
- Added sql debug support to virtual driver
- removed disable/enable methods in vacationdriver class
- Finally fixed GUI integration using the template

Version 20090912
- Reorganized code. All drivers move to lib/
- localcopy vs. keepcopy bugfix for DotForward/FTP driver
- enable/disable methods merged with setVacation()
- rewrote DotForward's parse() method
- Seperated forward and vacation in GUI

Version 20090909
- Rewrite of enable() method of the SQL driver
- Code refactoring and various cleanups

Version 20090908
- Backend -> Driver renamed and refactored
- Loading configuration file aligned with Roundcube's native function
- Finished SQL backend driver
- Added database schema for vacation tables
- Renamed config.php to config.inc.php
- Introducing /etc/postfixadmin/vacation.conf which gets generated using config.inc.php
- Added README.TXT with security notes and installation notes

Version 20090905
- Minor fix to setting enabled var
- Minor fix for parse() method in DotForward
- Changed default ftp-server to localhost
- Moved $config Array declaration from config to VacationBackend

Version 20090903

lib/virtual.class.php
<?php
/*
 * Virtual/SQL driver
 *
 * @package	plugins
 * @uses	rcube_plugin
 * @author	Jasper Slits <jaspersl at gmail dot com>
 * @version	1.9
 * @license GPL
 * @link	https://sourceforge.net/projects/rcubevacation/
 * @todo	See README.TXT
 * 
 * @moded 	2016-12-09 by LNDM for RC 1.2.3
 */

class Virtual extends VacationDriver {

	private $db, $domain, $domain_id, $goto = "";
	private $db_user;
	
	public function init() {
		// Use the DSN from db.inc.php or a dedicated DSN defined in config.ini

		if (empty($this->cfg['dsn'])) {
			$this->db = $this->rcmail->db;
			$dsn = rcube_db::parse_dsn($this->rcmail->config->get('db_dsnw'));
		} else {
			$this->db = new rcube_db($this->cfg['dsn'], '', FALSE);
			$this->db->db_connect('w');

			$this->db->set_debug((bool) $this->rcmail->config->get('sql_debug'));
			$dsn = rcube_db::parse_dsn($this->cfg['dsn']);
			$this->db->set_debug(true);

		}
		// Save username for error handling
		$this->db_user = $dsn['username'];

		if (isset($this->cfg['createvacationconf']) && $this->cfg['createvacationconf']) {

			$this->createVirtualConfig($dsn);
		}
	}

	/*
	 * @return Array Values for the form
	*/
	public function _get() {
		$vacArr = array("subject"=>"", "body"=>"");
		//   print_r($vacArr);
		$fwdArr = $this->virtual_alias();

		$sql = sprintf("SELECT subject,body,active FROM %s.vacation WHERE email='%s'",
				$this->cfg['dbase'], Q($this->user->data['username']));

		
		$res = $this->db->query($sql);
		if ($error = $this->db->is_error()) {
			raise_error(array('code' => 601, 'type' => 'db', 'file' => __FILE__,
						'message' => "Vacation plugin: query on {$this->cfg['dbase']}.vacation failed. Check DSN and verify that SELECT privileges on {$this->cfg['dbase']}.vacation are granted to user '{$this->db_user}'. <br/><br/>Error message:  " . $error), true, true);
		}



		if ($row = $this->db->fetch_assoc($res)) {
			$vacArr['body'] = utf8_encode($row['body']);
			$vacArr['subject'] = utf8_encode($row['subject']);
			$vacArr['enabled'] = ($row['active'] == 1) && ($fwdArr['enabled'] == 1);
		}


		return array_merge($fwdArr, $vacArr);
	}

	/*
	 * @return boolean True on succes, false on failure
	*/
	public function setVacation() {
		// If there is an existing entry in the vacation table, delete it.
		// This also triggers the cascading delete on the vacation_notification, but's ok for now.

		// We store since version 1.6 all data into one row.
		$aliasArr = array();

		// Sets class property
		$this->domain_id = $this->domainLookup();

		$sql = sprintf("UPDATE %s.vacation SET created=now(),active=0 WHERE email='%s'", $this->cfg['dbase'], Q($this->user->data['username']));


		$this->db->query($sql);

		$update = ($this->db->affected_rows() == 1);

		// Delete the alias for the vacation transport (Postfix)
		$sql = $this->translate($this->cfg['delete_query']);

		$this->db->query($sql);
		if ($error = $this->db->is_error()) {
			if (strpos($error, "no such field")) {
				$error = " Configure either domain_lookup_query or use %d in config.ini's delete_query rather than %i. <br/><br/>";
			}

			raise_error(array('code' => 601, 'type' => 'db', 'file' => __FILE__,
						'message' => "Vacation plugin: Error while saving records to {$this->cfg['dbase']}.vacation table. <br/><br/>" . $error
					), true, true);

		}


		// (Re)enable the vacation message and the vacation transport alias
		if ($this->enable && $this->body != "" && $this->subject != "") {

			if (!$update) {
				$sql = "INSERT INTO {$this->cfg['dbase']}.vacation  VALUES (?,?,?,'',?,NOW(),1)";
			} else {
				$sql = "UPDATE {$this->cfg['dbase']}.vacation SET email=?,subject=?,body=?,domain=?,active=1 WHERE email=?";
			}


			$this->db->query($sql, Q($this->user->data['username']), $this->subject, $this->body, $this->domain,Q($this->user->data['username']));
			if ($error = $this->db->is_error()) {
				if (strpos($error, "no such field")) {
					$error = " Configure either domain_lookup_query or use \%d in config.ini's insert_query rather than \%i<br/><br/>";
				}

				raise_error(array('code' => 601, 'type' => 'db', 'file' => __FILE__,
							'message' => "Vacation plugin: Error while saving records to {$this->cfg['dbase']}.vacation table. <br/><br/>" . $error
						), true, true);
			}
			$aliasArr[] = '%g';
		}


		// Keep a copy of the mail if explicitly asked for or when using vacation
		$always = (isset($this->cfg['always_keep_copy']) && $this->cfg['always_keep_copy']);
		if ($this->keepcopy || in_array('%g', $aliasArr) || $always) {
			$aliasArr[] = '%e';
		}

		// Set a forward
		if ($this->forward != null) {
			$aliasArr[] = '%f';
		}

		// Aliases are re-created if $sqlArr is not empty.
		$sql = $this->translate($this->cfg['delete_query']);
		$this->db->query($sql);

		// One row to store all aliases
		if (!empty($aliasArr)) {

			$alias = join(",", $aliasArr);
			$sql = str_replace('%g', $alias, $this->cfg['insert_query']);
			$sql = $this->translate($sql);

			$this->db->query($sql);
			if ($error = $this->db->is_error()) {
				raise_error(array('code' => 601, 'type' => 'db', 'file' => __FILE__,
							'message' => "Vacation plugin: Error while executing {$this->cfg['insert_query']} <br/><br/>" . $error
						), true, true);
			}
		}
		return true;
	}

	/*
	 * @return string SQL query with substituted parameters
	*/
	private function translate($query) {
		return str_replace(array('%e', '%d', '%i', '%g', '%f', '%m'),
				array($this->user->data['username'], $this->domain, $this->domain_id,
					Q($this->user->data['username']) . "@" . $this->cfg['transport'], $this->forward, $this->cfg['dbase']), $query);
	}

// Sets %i. Lookup the domain_id based on the domainname. Returns the domainname if the query is empty
	private function domainLookup() {
		// Sets the domain
		list($username, $this->domain) = explode("@", $this->user->get_username());
		if (!empty($this->cfg['domain_lookup_query'])) {
			$res = $this->db->query($this->translate($this->cfg['domain_lookup_query']));

			if (!$row= $this->db->fetch_array($res)) {
				raise_error(array('code' => 601, 'type' => 'db', 'file' => __FILE__,
							'message' => "Vacation plugin: domain_lookup_query did not return any row. Check config.ini <br/><br/>" . $this->db->is_error()
						), true, true);

			}
			return $row[0];
		} else {
			return $this->domain;
		}
	}

	/*Creates configuration file for vacation.pl
	 *
	 * @param array dsn
	 * @return void
	*/
	private function createVirtualConfig(array $dsn) {

		$virtual_config = "/etc/postfixadmin/";
		if (!is_writeable($virtual_config)) {
			raise_error(array('code' => 601, 'type' => 'php', 'file' => __FILE__,
						'message' => "Vacation plugin: Cannot create {$virtual_config}vacation.conf . Check permissions.<br/><br/>"
					), true, true);
		}

		// Fix for vacation.pl
		if ($dsn['phptype'] == 'pgsql') {
			$dsn['phptype'] = 'Pg';
		}

		$virtual_config .= "vacation.conf";
		// Only recreate vacation.conf if config.ini has been modified since
		if (!file_exists($virtual_config) || (filemtime("plugins/vacation/config.ini") > filemtime($virtual_config))) {
			$config = sprintf("
		\$db_type = '%s';
		\$db_username = '%s';
		\$db_password = '%s';
		\$db_name     = '%s';
		\$vacation_domain = '%s';", $dsn['phptype'], $dsn['username'], $dsn['password'], $this->cfg['dbase'], $this->cfg['transport']);
			file_put_contents($virtual_config, $config);
		}
	}

	/*
			Retrieves the localcopy and/or forward settings.
		* @return array with virtual aliases
	*/
	private function virtual_alias() {
		$forward = "";
		$enabled = false;
		$goto = Q($this->user->data['username']) . "@" . $this->cfg['transport'];

		// Backwards compatiblity. Since >=1.6 this is no longer needed
		$sql = str_replace("='%g'", "<>''", $this->cfg['select_query']);

		$res = $this->db->query($this->translate($sql));

		$rows = array();

		while (list($row) = $this->db->fetch_array($res)) {

			// Postfix accepts multiple aliases on 1 row as well as an alias per row
			if (strpos($row, ",") !== false) {
				$rows = explode(",", $row);
			} else {
				$rows[] = $row;
			}
		}



		foreach ($rows as $row) {
			// Source = destination means keep a local copy
			if ($row == $this->user->data['username']) {
				$keepcopy = true;
			} else {
				// Neither keepcopy or postfix transport means it's an a forward address
				if ($row == $goto) {
					$enabled = true;
				} else {
					// Support multi forwarding addresses
					$forward .= $row . ",";
				}
			}

		}
		// Substr removes any trailing comma
		return array("forward"=>substr($forward, 0,  - 1), "keepcopy"=>$keepcopy, "enabled"=>$enabled);
	}

// Destroy the database connection of our temporary database connection
	public function __destruct() {
		if (!empty($this->cfg['dsn']) && is_resource($this->db)) {
			$this->db = null;
		}
	}
}

?>


lib/vacationdriver.class.php
<?php
/*
 * VacationDriver base class
 *
 * @package	plugins
 * @uses	rcube_plugin
 * @author	Jasper Slits <jaspersl at gmail dot com>
 * @version	1.9
 * @license     GPL
 * @link	https://sourceforge.net/projects/rcubevacation/
 * @todo	See README.TXT
 */

abstract class VacationDriver {
	protected $cfg,$dotforward = array();
	protected $rcmail,$user,$forward,$body,$subject,$aliases = "";
	protected $enable,$keepcopy = false;

	abstract public function _get();
	abstract public function init();
	abstract protected function setVacation();

	// Provide easy access for the drivers to frequently used objects
	public function __construct() {
		$this->rcmail = rcmail::get_instance();
		$this->user = $this->rcmail->user;
		$this->identity = $this->user->get_identity();
		$child_class = strtolower( get_class($this));
	}

        /*
        
         */
	public final function setIniConfig(array $inicfg)
	{
		$this->cfg = $inicfg;
	}


	public final function setDotForwardConfig($child_class,$config) {
		// forward settings are shared by ftp,sshftp and setuid driver. 
		if (in_array($child_class ,array('ftp','sshftp','setuid'))) {
			$this->dotforward = $config;
		}
	}

	// Helper method for the template to determine if user is allowed to enter aliases
	public final function useAliases()
	{

		return (isset($this->dotforward['alias_identities']) && $this->dotforward['alias_identities']);
	}

public function loadDefaults() {
    // Load default subject and body.

    if (empty($this->cfg['body'])) return false;


    $file = "plugins/vacation/" . $this->cfg['body'];
    

    if (is_readable($file)) {
        $defaults = array('subject'=>$this->cfg['subject']);
        $defaults['body'] = file_get_contents($file);
        return $defaults;
    } else {
        raise_error(array('code' => 601, 'type' => 'php', 'file' => __FILE__,
                    'message' => sprintf("Vacation plugin: s cannot be opened", $file)
                ), true, true);
    }
}
	
	

	// This method will be used from vacation.js as an JSON/Ajax call or directly
	public final function vacation_aliases($method=null)
	{
		$aliases = "";
		$identities = $this->user->list_identities();
		// Strip off the default identity, no need to alias that.
		array_shift($identities);
	  
		

		foreach($identities as $identity)
		{
			// Strip domainname off. /usr/bin/vacation only deals with system users
			
			$aliases.=array_shift(explode("@",$identity['email'])).",";
		}

		$str = substr($aliases,0,-1);

		// We use this method in both ftp.class.php and as Ajax callback

		if (empty($identities))
		{
			// No identities found.
//			$str = "To use aliases, add more identities.";
		}

		if ($method != null)
		{
			return $str;
		} else {
			// Calls the alias_callback as defined in vacation.js
			$this->rcmail->output->command('plugin.alias_callback',  array('aliases'=>$str));
		}
	}


	/*
	 * @return boolean True on succes, false on failure
	 */
	final public function save() {
		$this->enable = (NULL != get_input_value('_vacation_enabled', RCUBE_INPUT_POST));
		$this->subject = utf8_decode(get_input_value('_vacation_subject', RCUBE_INPUT_POST));
		$this->body = utf8_decode(get_input_value('_vacation_body', RCUBE_INPUT_POST));
		$this->keepcopy = (NULL != get_input_value('_vacation_keepcopy', RCUBE_INPUT_POST));
		$this->forward = get_input_value('_vacation_forward', RCUBE_INPUT_POST);
		$this->aliases = get_input_value('_vacation_aliases', RCUBE_INPUT_POST);

		// This method performs the actual work
		return $this->setVacation();
	}

	final public function getActionText()
	{
		if ($this->enable && empty($this->forward)) { return "enabled_and_no_forward"; };
		if ($this->enable && !empty($this->forward)) {  return "enabled_and_forward"; };
		if (! $this->enable && !empty($this->forward)) { return "disabled_and_forward"; };
		if (! $this->enable && empty($this->forward)) { return "disabled_and_no_forward"; };
		
	}
}?>


vacation.php
<?php
/*
 * Vacation plugin that adds a new tab to the settings section
 * to enable forward / out of office replies.
 *
 * @package	plugins
 * @uses	rcube_plugin
 * @author	Jasper Slits <jaspersl@gmail.com>
 * @version	1.9
 * @license     GPL
 * @link	https://sourceforge.net/projects/rcubevacation/
 * @todo	See README.TXT
 *
 * @moded	2016-12-09 by LNDM for RC 1.2.3

*/

// Load required dependencies
require 'lib/vacationdriver.class.php';
require 'lib/dotforward.class.php';
require 'lib/vacationfactory.class.php';
require 'lib/VacationConfig.class.php';

class vacation extends rcube_plugin {
	
	public $task = 'settings';
	private $v = "";
	private $inicfg = "";
	private $enableVacationTab = true;
	private $vcObject;
	
	public function init() {
		$this->add_texts('localization/', array('vacation'));
		$this->load_config();
		
		$this->inicfg = $this->readIniConfig();

		

		// Don't proceed if the current host does not support vacation
		if (!$this->enableVacationTab) {
			return false;
		}

		$this->v = VacationDriverFactory::Create($this->inicfg['driver']);

		$this->v->setIniConfig($this->inicfg);
		$this->register_action('plugin.vacation', array($this, 'vacation_init'));
		$this->register_action('plugin.vacation-save', array($this, 'vacation_save'));
		$this->register_handler('plugin.vacation_form', array($this, 'vacation_form'));
		// The vacation_aliases method is defined in vacationdriver.class.php so use $this->v here
		$this->register_action('plugin.vacation_aliases', array($this->v, 'vacation_aliases'));
		$this->include_script('vacation.js');
		$this->include_stylesheet('skins/default/vacation.css');
		$this->rcmail = rcmail::get_instance();
		$this->user = $this->rcmail->user;
		$this->identity = $this->user->get_identity();
		
		// forward settings are shared by ftp,sshftp and setuid driver.
		$this->v->setDotForwardConfig($this->inicfg['driver'],$this->vcObject->getDotForwardCfg());
	}
	
	public function vacation_init() {
		$this->add_texts('localization/', array('vacation'));
		$rcmail = rcmail::get_instance();
		$rcmail->output->set_pagetitle($this->gettext('autoresponder'));
		//Load template
		$rcmail->output->send('vacation.vacation');
	}
	
	public function vacation_save() {
		$rcmail = rcmail::get_instance();

		// Initialize the driver
		$this->v->init();

		if ($this->v->save()) {
//          $this->v->getActionText() Dummy for now
			$rcmail->output->show_message($this->gettext("success_changed"), 'confirmation');
		} else {
			$rcmail->output->show_message($this->gettext("failed"), 'error');
		}
		$this->vacation_init();
	}

	// Parse config.ini and get configuration for current host
	private function readIniConfig() {
		$this->vcObject = new VacationConfig();
		$this->vcObject->setCurrentHost($_SESSION['imap_host']);
		$config = $this->vcObject->getCurrentConfig();

		if (false !== ($errorStr = $this->vcObject->hasError())) {
			raise_error(array('code' => 601, 'type' => 'php', 'file' => __FILE__,
						'message' => sprintf("Vacation plugin: %s", $errorStr)), true, true);
		}
		$this->enableVacationTab = $this->vcObject->hasVacationEnabled();

		return $config;
	}
	
	public function vacation_form() {
		$rcmail = rcmail::get_instance();
		// Initialize the driver
		$this->v->init();
		$settings = $this->v->_get();

		// Load default body & subject if present.
		if (empty($settings['subject']) && $defaults = $this->v->loadDefaults()) {
			$settings['subject'] = $defaults['subject'];
			$settings['body'] = $defaults['body'];
		}

		$rcmail->output->set_env('product_name', $rcmail->config->get('product_name'));
		// return the complete edit form as table
		$out = '<fieldset><legend>' . $this->gettext('outofoffice') . ' ::: ' . $rcmail->user->data['username'] . '</legend>' . "\n";
		// show autoresponder properties

		// Auto-reply enabled
		$field_id = 'vacation_enabled';
		$input_autoresponderactive = new html_checkbox(array('name' => '_vacation_enabled', 'id' => $field_id, 'value' => 1));
		$out .= sprintf("<p><label for=\"%s\">%s</label>&nbsp;%s</p>\n",
				$field_id,
				rep_specialchars_output($this->gettext('autoreply')),
				$input_autoresponderactive->show($settings['enabled']));

		// Subject
		$field_id = 'vacation_subject';
		$input_autorespondersubject = new html_inputfield(array('name' => '_vacation_subject', 'id' => $field_id, 'size' => 90));
		$out .= sprintf("<p><label for=\"%s\">%s</label>&nbsp;%s</p>\n",
				$field_id,
				rep_specialchars_output($this->gettext('autoreplysubject')),
				$input_autorespondersubject->show($settings['subject']));

		// Out of office body
		$field_id = 'vacation_body';
		$input_autoresponderbody = new html_textarea(array('name' => '_vacation_body', 'id' => $field_id, 'cols' => 70, 'rows' => 15));
		$out .= sprintf("<p><label style=\"vertical-align:top;\" for=\"%s\">%s</label>&nbsp;%s</p>\n",
				$field_id,
				rep_specialchars_output($this->gettext('autoreplymessage')),
				$input_autoresponderbody->show($settings['body']));

		/* We only use aliases for .forward and only if it's enabled in the config*/
		if ($this->v->useAliases()) {
		$size = 0;

		// If there are no multiple identities, hide the button and add increase the size of the textfield
		$hasMultipleIdentities = $this->v->vacation_aliases('buttoncheck');
		if ($hasMultipleIdentities == '') $size = 15;

				$field_id = 'vacation_aliases';
				$input_autoresponderalias = new html_inputfield(array('name' => '_vacation_aliases', 'id' => $field_id, 'size' => 75+$size));
				$out .= '<p>' . $this->gettext('separate_alias') . '</p>';

			// Inputfield with button
			$out .= sprintf('<p><label for="%s">%s</label>&nbsp;%s
			  ', $field_id, rep_specialchars_output($this->gettext('aliases')),
				  $input_autoresponderalias->show($settings['aliases']));
			if ($hasMultipleIdentities!='')
				$out .= sprintf('<input type="button" id="aliaslink" class="button" value="%s"/>',
			  rep_specialchars_output($this->gettext('aliasesbutton')));
			$out .= "</p>";

		}
		$out .= '</fieldset><br /><fieldset><legend>' . $this->gettext('forward') . '</legend>';

			// Keep a local copy of the mail
			$field_id = 'vacation_keepcopy';
			$input_localcopy = new html_checkbox(array('name' => '_vacation_keepcopy', 'id' => $field_id, 'value' => 1));
			$out .= sprintf("<p><label for=\"%s\">%s</label>&nbsp;%s</p>\n",
					$field_id,
					rep_specialchars_output($this->gettext('keepcopy')),
					$input_localcopy->show($settings['keepcopy']));
			


		// Information on the forward in a seperate fieldset.
		if (! isset($this->inicfg['disable_forward']) || ( isset($this->inicfg['disable_forward']) && $this->inicfg['disable_forward']==false))
		{
						$out .= '<p>' . $this->gettext('separate_forward') . '</p>';

			// Forward mail to another account
			$field_id = 'vacation_forward';
			$input_autoresponderforward = new html_inputfield(array('name' => '_vacation_forward', 'id' => $field_id, 'size' => 90));
			$out .= sprintf("<p><label for=\"%s\">%s</label>&nbsp;%s</p>\n",
					$field_id,
					rep_specialchars_output($this->gettext('forwardingaddresses')),
					$input_autoresponderforward->show($settings['forward']));

			
		}
				$out .= "</fieldset><br />\n";
		$rcmail->output->add_gui_object('vacationform', 'vacation-form');
		return $out;
	}
}

?>
